source/opt/optimizer.hpp

```diff
@@ -53,99 +73,98 @@ class Optimizer {
-  // the module. E.g., an integer spec constant defined with OpSpecConstantOp
-  // instruction can be folded if its value won't change later. This pass will
-  // replace the original OpSpecContantOp instruction with an OpConstant
-  // instruction. When folding composite spec constants, new instructions
-  // may be inserted to define the components of the composite constant first,
-  // then the original spec constants will be replaced by OpConstantComposite
-  // instructions.
-  //
-  // There are some operations not supported yet:
-  //   OpSConvert, OpFConvert, OpQuantizeToF16 and
-  //   all the operations under Kernel capability.
-  // TODO(qining): Add support for the operations listed above.
-  Optimizer& RegisterFoldSpecConstantOpAndCompositePass();
-
-  // Registers a unify-constant pass to this optimizer.
-  // A unify-constant pass de-duplicates the constants. Constants with the exact
-  // same value and identical form will be unified and only one constant will
-  // be kept for each unique pair of type and value.
-  // There are several cases not handled by this pass:
-  //  1) Constants defined by OpConstantNull instructions (null constants) and
-  //  constants defined by OpConstantFalse, OpConstant or OpConstantComposite
-  //  with value 0 (zero-valued normal constants) are not considered equivalent.
-  //  So null constants won't be used to replace zero-valued normal constants,
-  //  vice versa.
-  //  2) Whenever there are decorations to the constant's result id id, the
-  //  constant won't be handled, which means, it won't be used to replace any
-  //  other constants, neither can other constants replace it.
-  //  3) NaN in float point format with different bit patterns are not unified.
-  Optimizer& RegisterUnifyConstantPass();
-
-  // Registers a eliminate-dead-constant pass to this optimizer.
-  // A eliminate-dead-constant pass removes dead constants, including normal
-  // contants defined by OpConstant, OpConstantComposite, OpConstantTrue, or
-  // OpConstantFalse and spec constants defined by OpSpecConstant,
-  // OpSpecConstantComposite, OpSpecConstantTrue, OpSpecConstantFalse or
-  // OpSpecConstantOp.
-  Optimizer& RegisterEliminateDeadConstantPass();
-
-  // Clears all passes previously registered.
-  Optimizer& ClearRegisteredPasses();
+  // Registers the given |pass| to this optimizer. Passes will be run in the
+  // exact order of registration.
+  Optimizer& RegisterPass(PassToken&& pass);
+
+  // Optimizes the given SPIR-V module |original_binary| and writes the
+  // optimized binary into |optimized_binary|.
+  // Returns true on successful optimization, whether or not the module is
+  // modified. Returns false if errors occur when processing |original_binary|
+  // using any of the registered passes. In that case, no further passes are
+  // excuted and the contents in |optimized_binary| may be invalid.
+  bool Run(const uint32_t* original_binary, size_t original_binary_size,
```

Comment:
You should say that it's ok for original_binary to alias the start of the optimized_binary vector.